fastapi/dependencies/utils.py

```diff
@@ -536,7 +544,11 @@ async def solve_dependencies(
         if sub_dependant.use_cache and sub_dependant.cache_key in dependency_cache:
             solved = dependency_cache[sub_dependant.cache_key]
         elif is_gen_callable(call) or is_async_gen_callable(call):
-            stack = request.scope.get("fastapi_astack")
+            if sub_dependant.exit_before_response:
+                stack = request.scope.get("fastapi_astack_before_response")
+            else:
+                stack = request.scope.get("fastapi_astack")
+
```

Comment:
```
key = 'fastapi_astack_before_response' if sub_dependant.exit_before_response else 'fastapi_astack'
stack = request.scope.get(key)
```

Reply:
usually, I do it like you commented, but this is very sesitive matter

when doing stuff on sensitive matter, I prefer to be explicit, it is easier to see what is going on

i will fix if this will ever be a candidate to merge :)